### DISCUSSION TITLE
"Implicit" Attribute for generic parameters

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#1103] [General]
Here is the lightbulb! no new syntax, no breaking change. no compiler option.

introduce "Implicit" attribute for generic parameters so that compiler will automatically infer generic type arguments when method is called.

```
void Foo<TSource, [Implicit] T>(TSource x) where TSource : IList<T>
{
}
```

this will let compiler know that we intend to infer `T` implicitly. because its already some where hidden in input! (in this case `IList<T>`) so as an example, following call becomes valid

    Foo(new List<int>());

This special attribute must also cause compile time error if marked generic parameter is not found anywhere in input, for example

```
void Foo<[Implicit] T>() // Error: no usage of T found in method signature.
{
}
```

**Comment by orthoxerox** [comment]
I think implicit generic parameters are being prototyped in the concepts fork.

cc @MattWindsor91

**Comment by MattWindsor91** [comment]
@orthoxerox Sadly only for concept parameters.

Having it work for normal type parameters would be a logical extension and make the concepts version feel more orthogonal and less weird, but I genuinely don’t know how much would this be useful outside of concepts as I haven’t looked into it.

**Comment by gafter** [comment]
@agocke is investigating extending type inference to take constraints into account.

**Comment by alrz** [comment]
>extending type inference

What is the scope of this feature? would we be able to consider other improvements for type inference? for instance: https://github.com/dotnet/csharplang/issues/92, https://github.com/dotnet/csharplang/issues/877, https://github.com/dotnet/csharplang/issues/281, etc
